Avoid panics in LSP store path handling

CLoaKY233 opened the pull request:
Release Notes:

- Fixed incorrect journal paths handling

CLoaKY233 committed:
Avoid panics in LSP store path handling

Use log::warn! when URI conversion fails during buffer unregistration.
Replace unwrap() on Uri::from_file_path and as_local_mut() with Option-
friendly handling (ok()? and ?), returning early instead of panicking.

CLoaKY233 committed:
Fix buffer unregistration error logging format

SomeoneToIgnore reviewed (changes requested):
> The panic was caused by a `debug_panic!` which has been replaced with error handling.

This is wrong for release builds:

https://github.com/zed-industries/zed/blob/21f73d9c02681152019ed5703ce8808c841fcbbe/crates/util/src/util.rs#L42-L52

and 

> Invalid file path provided to LSP request

panic message is used elsewhere: https://github.com/zed-industries/zed/blob/21f73d9c02681152019ed5703ce8808c841fcbbe/crates/project/src/lsp_command.rs#L53-L59

I have not found direct usages of `file_path_to_lsp_url` calling `.unwrap` or analogues, so apparently we need to search further up the stack.

So:

* the issue does not seem to be in any of the code the PR edits?
It would be good to get a sequence of actions to reproduce the bug — in the issue description; and to get a test for this into the PR.

* `debug_panic` change seems wrong, hence needs to be reverted.
The rest of the changes seems nice — I can merge them alone (but that won't fix the issue, it seems?) or wait for a better fix, if you prefer.

CLoaKY233 commented:
> This is wrong for release builds:
> 
> https://github.com/zed-industries/zed/blob/21f73d9c02681152019ed5703ce8808c841fcbbe/crates/util/src/util.rs#L42-L52
> 
> and
> 
> > Invalid file path provided to LSP request
> 
> panic message is used elsewhere:
> 

I  missed to mention the first unwrap call, post fixing that, the debug panic was causing crashes, 
`crates/project/src/lsp_store.rs` 

line `7654` 
```rust
        let buffer = buffer.read(cx);
        let file = File::from_dyn(buffer.file())?;
        let abs_path = file.as_local()?.abs_path(cx);
        let uri = lsp::Uri::from_file_path(abs_path).unwrap(); //<----------- This 
 ```
 
  
  ```log
thread 'main' panicked at crates/project/src/lsp_store.rs:7654:54:
called `Result::unwrap()` on an `Err` value: ()

```

So I guess we can revert the debug_panic! change, but the second change, does fix the crash in release builds. 

--- 
although what I do not understand is it creates the invalid path I have provided in the zed's directory instead, 
Eg, specifying a journal path `NONEXIST/xyz/abc/Journal` 

it creates the respective directories in the zed's directory instead, 
`/zed/NONEXT222/home/cloaky/Journal/journal/2025/11/`

and upon replacing the `unwrap()` with `ok()?` 

Error logs are : 
```log
2025-11-11T13:54:24+05:30 ERROR [crates/project/src/environment.rs:227] stat ""

Caused by:
    No such file or directory (os error 2)
2025-11-11T13:54:24+05:30 ERROR [project::git_store] failed to get working directory environment for repository ""
2025-11-11T13:54:24+05:30 INFO  [worktree] inserting parent git repo for this worktree: "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md"
2025-11-11T13:54:24+05:30 ERROR [crates/project/src/lsp_store.rs:2425] Invalid file path provided to LSP request: "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md"
2025-11-11T13:54:24+05:30 INFO  [worktree] inserting parent git repo for this worktree: "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md"
2025-11-11T13:54:24+05:30 ERROR [crates/project/src/lsp_store.rs:4336] Invalid file path provided to LSP request: "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md"
2025-11-11T13:54:24+05:30 ERROR [crates/project/src/lsp_store.rs:2425] Invalid file path provided to LSP request: "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md"
2025-11-11T13:54:24+05:30 ERROR [crates/project/src/lsp_store.rs:2425] Invalid file path provided to LSP request: "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md"
2025-11-11T13:54:24+05:30 ERROR [project::lsp_store] "NONEXT222/home/cloaky/Journal/journal/2025/11/11.md" is not parseable as an URI
<disabled>
2025-11-11T13:54:29+05:30 ERROR [crates/gpui/src/window.rs:1131] window not found
2025-11-11T13:54:29+05:30 ERROR [crates/gpui/src/window.rs:1138] window not found



```
--- 
If its intentional that it gets created in the zed dir, we can maybe check the zed path as fallback when this function runs and path is not found? 
Would this be a good approach ? @SomeoneToIgnore

CLoaKY233 committed:
Merge branch 'zed-industries:main' into Journal-crash-fix

CLoaKY233 committed:
Use debug_panic! when URI conversion fails [revert change]

CLoaKY233 commented:
`journal/src/journal.rs:175-179`
There's the issue ! 

```rs 
// Excerpt from: fn journal_dir
fn journal_dir(path: &str) -> Option<PathBuf> {
    shellexpand::full(path) //TODO handle this better
        .ok()
        .map(|dir| Path::new(&dir.to_string()).to_path_buf().join("journal"))
}
```
```log
2025-11-11T15:27:37+05:30 INFO  [journal] journal_dir called with path: somedirectory/home/cloaky/Journal
2025-11-11T15:27:37+05:30 INFO  [journal] shellexpand::full result: Ok("somedirectory/home/cloaky/Journal")
2025-11-11T15:27:37+05:30 INFO  [journal] Expanded directory: somedirectory/home/cloaky/Journal
2025-11-11T15:27:37+05:30 INFO  [journal] PathBuf created: "somedirectory/home/cloaky/Journal"
2025-11-11T15:27:37+05:30 INFO  [journal] Final journal path: "somedirectory/home/cloaky/Journal/journal"
```

Does not return an absolute path,

CLoaKY233 committed:
Make journal path absolute

- Journal path expansion now returns an absolute path: expand the input,
treat absolute paths as-is, and warn/fallback to the user's home
directory if given a relative path. Added unit tests for absolute,
tilde, relative (fallback), and Windows-style paths.

CLoaKY233 committed:
Merge branch 'main' into Journal-crash-fix

SomeoneToIgnore reviewed (commented):
Thank you for the context, now it gets quite clear indeed.
Let's merge after a few small tweaks.

CLoaKY233 committed:
Merge branch 'zed-industries:main' into Journal-crash-fix

CLoaKY233 committed:
Match paths for journal tests and use log_err for URI

CLoaKY233 commented on crates/project/src/lsp_store.rs:
```diff
@@ -7654,14 +7654,13 @@ impl LspStore {
             .ok()
             .with_context(|| format!("Failed to convert path to URI: {}", abs_path.display()))
-            .unwrap();
+            .ok()?;
```
Changed to `.log_err()?` to make the error visible. :D

CLoaKY233 committed:
Restrict absolute path test to Unix

CLoaKY233 committed:
Clarify path fallback behavior in docs

SomeoneToIgnore reviewed (approved):
Thank you.
